### PR TITLE
NCS-357 Add shareholder endpoints

### PR DESCRIPTION
**Pull request by AlfredBwomezi**
NCS-357 Add endpoints to service shareholder details' requests.

- Endpoint for shareholder count.
- Endpoint for shareholder (details) list.

Linked to [NCS-87](https://companieshouse.atlassian.net/browse/NCS-87).

### CODE REVIEW

File: src/main/java/uk/gov/ch/controller/ShareholderController.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OracleQueryApplication.APPLICATION_NAME_SPACE);
+
+    @Autowired
+    private ShareholderService shareholderService;
+
+    @GetMapping("/company/{companyNumber}/shareholders/count")
+    public ResponseEntity<Integer> getShareholdersCount(@PathVariable String companyNumber) {
+
+        LOGGER.info("Calling service to retrieve shareholder count for corporate body ID " + companyNumber);
```
Comment (markpit) [inline src/main/java/uk/gov/ch/controller/ShareholderController.java:29]: log message still references corporate body ID - could you update to be "company number"

File: src/main/java/uk/gov/ch/controller/ShareholderController.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        int response = shareholderService.getShareholderCount(companyNumber);
+
+        LOGGER.info("Returning shareholder count ("+ response +") for company " + companyNumber);
+        return ResponseEntity.status(HttpStatus.OK).body(response);
+    }
+
+    @GetMapping("/company/{companyNumber}/shareholders")
+    public ResponseEntity<List<Shareholder>> getShareholders(@PathVariable String companyNumber) {
+        
+        LOGGER.info("Calling service to retrieve the list of shareholders for corporate body ID " + companyNumber);
```
Comment (markpit) [inline src/main/java/uk/gov/ch/controller/ShareholderController.java:39]: same here for the log message referencing "corporate body ID"

File: src/main/java/uk/gov/ch/controller/ShareholderController.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        return ResponseEntity.status(HttpStatus.OK).body(response);
+    }
+
+    @GetMapping("/company/{companyNumber}/shareholders")
+    public ResponseEntity<List<Shareholder>> getShareholders(@PathVariable String companyNumber) {
+        
+        LOGGER.info("Calling service to retrieve the list of shareholders for company number " + companyNumber);
+        List<Shareholder> response = shareholderService.getShareholders(companyNumber);
+
+        LOGGER.info("Returning "+ response.size() +"shareholders for company " + companyNumber);
```
Comment (markpit) [inline src/main/java/uk/gov/ch/controller/ShareholderController.java:42]: response.size() +"shareholders. needs a space before shareholders eg " shareholders"

File: src/main/java/uk/gov/ch/dao/ShareholderDao.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        return list;
+    }
+
+    public int getShareholderCount(String incorporationNumber) {
+        int count;
+        count = jdbcTemplate.queryForObject(SHAREHOLDER_COUNT_SQL, Integer.class, incorporationNumber);
+        
+        LOGGER.info("Returned shareholders count of" + count + " shareholders from SHAREHOLDER tables.");
```
Comment (markpit) [inline src/main/java/uk/gov/ch/dao/ShareholderDao.java:58]: need space after of" in log message eg . of "

File: src/main/java/uk/gov/ch/dao/ShareholderDao.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+            LOGGER.info("Returned " + list.size() + " shareholders from SHAREHOLDER_ELECTED tables.");
+        }
+
+        return list;
+    }
+
+    public int getShareholderCount(String incorporationNumber) {
+        int count;
+        count = jdbcTemplate.queryForObject(SHAREHOLDER_COUNT_SQL, Integer.class, incorporationNumber);
```
Comment (markpit) [inline src/main/java/uk/gov/ch/dao/ShareholderDao.java:56]: count can be declared and assigned on one line here eg
int count = jdbcTemplate.queryForObject(.....

File: src/main/java/uk/gov/ch/dao/ShareholderDao.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        
+        LOGGER.info("Returned shareholders count of" + count + " shareholders from SHAREHOLDER tables.");
+        
+        if (count == 0) {
+            count = jdbcTemplate.queryForObject(SHAREHOLDER_ELECTED_COUNT_SQL, Integer.class, incorporationNumber);
+            
+            LOGGER.info("Returned shareholders count of" + count + " shareholders from SHAREHOLDER_ELECTED tables.");
```
Comment (markpit) [inline src/main/java/uk/gov/ch/dao/ShareholderDao.java:63]: space needed after of". (before quote), eg of "

File: src/test/java/uk/gov/ch/controller/ShareholderControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import org.springframework.http.ResponseEntity;
+
+import uk.gov.ch.model.Shareholder;
+import uk.gov.ch.service.ShareholderService;
+
+@ExtendWith(MockitoExtension.class)
+class ShareholderControllerTest {
+
+    @InjectMocks
+    ShareholderController controller;
```
Comment (markpit) [inline src/test/java/uk/gov/ch/controller/ShareholderControllerTest.java:25]: the controller and service can be marked as private

File: src/test/java/uk/gov/ch/dao/ShareholderDaoTest.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import uk.gov.ch.model.Shareholder;
+
+@ExtendWith(MockitoExtension.class)
+class ShareholderDaoTest {
+
+    @Mock
+    private JdbcTemplate jdbcTemplate;
+
+    @InjectMocks
+    ShareholderDao dao;
```
Comment (markpit) [inline src/test/java/uk/gov/ch/dao/ShareholderDaoTest.java:33]: dao can be privatr

File: src/main/java/uk/gov/ch/dao/ShareholderDao.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+@Repository
+public class ShareholderDao {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OracleQueryApplication.APPLICATION_NAME_SPACE);
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    static final String SHAREHOLDER_COUNT_SQL = "SELECT COUNT(*) FROM shareholder sh INNER JOIN shareholding shd ON sh.shareholder_id = shd.shareholding_id JOIN corporate_body cb on cb.corporate_body_id = shd.corporate_body_id WHERE cb.incorporation_number = ?";
+    static final String SHAREHOLDER_ELECTED_COUNT_SQL = "SELECT COUNT(*) FROM shareholder_elected sh JOIN sholdingelc_sholderelc_link shl ON sh.shareholder_elected_id = shl.shareholder_elected_id  INNER JOIN shareholding_elected shd ON shd.shareholding_elected_id = shl.shareholding_elected_id JOIN corporate_body cb on cb.corporate_body_id = shd.corporate_body_id WHERE cb.incorporation_number = ?";
```
Comment (markpit) [inline src/main/java/uk/gov/ch/dao/ShareholderDao.java:27]: i'd add a blank line between the 2 sqls line 26/27 just to make it a bit easier to read. You've done it for the others, looks like this one got missed.

**Comment by markpit** [comment]
Looks good to me, I'll hold off approving until we hear back from the DBA review of the queries

**Comment by AlfredBwomezi** [comment]
DBA is [OK](https://companieshouse.slack.com/archives/C11FMAQBT/p1623236753018600) with the scripts.

**Review by markpit — APPROVED** [review]

**Pull request merged**